NumPy scalars to ctypes conversion support

aldanor opened the pull request:
This adds direct converters that don't require `detail::type_info` to operate + support for passing NumPy structured scalars back to C++ without instance tracking and Python object creation overhead.

See #451 for discussion.

Note that this also disables registering numpy dtypes more than once (since then there would exist multiple conflicting direct converters); if we want that functionality back (do we?), it could probably be fixed, e.g. `npy_format_descriptor<>` could hold a pointer to the callback so that it can be later removed from the list if needed.

@wjakob I may have overlooked some details, so comments are welcome :)

wjakob commented on include/pybind11/cast.h:
```diff
@@ -208,13 +208,21 @@ class type_caster_generic {
                 if (load(temp, false))
                     return true;
             }
+            auto& direct = get_internals().direct_conversions;
```
Let's not call ``get_internals()`` over and over again in this function, we can cache it.

wjakob commented on include/pybind11/cast.h:
```diff
@@ -157,7 +157,8 @@ inline void keep_alive_impl(handle nurse, handle patient);
     PYBIND11_NOINLINE type_caster_generic(const std::type_info &type_info)
-     : typeinfo(get_type_info(type_info, false)) { }
+     : typeinfo(get_type_info(type_info, false)),
+       direct_conversions(get_internals().direct_conversions[std::type_index(type_info)]) { }
```
Not so excited about the thought of another global hash map lookup here -- these are kind of expensive as was noticed in another context. I'll have to think a bit about this.

aldanor replied:
Well, otherwise you'll do a lookup on each `load()`, like we do with a whole bunch of other stuff, so it's no worse. It's probably possible to cache it somewhere else though, but it's not immediately obvious.

aldanor commented:
@wjakob What we can also do, I guess, is stick `direct_conversions` and `registered_types_cpp` in one map, with struct values that contain (optional) pointers to both types of values; this is possible because both are indexed by `type_index`. This way you'll only do one lookup.

aldanor commented:
One thing I don't fully understand is why `type_caster_generic` is not templated on the type. There's only 6 occurrences where it's used in the code, in all of which the type is known exactly. If it was templated, you could precache all kinds of things, e.g. registered_cpp_type for casts, direct converters for loads, etc.

wjakob commented:
Hi @aldanor,

to avoid code bloat, it's essential to keep as much code as possible out of templates. Even if a function is only used in one or two places, it will be costly to implement it in a template if that means that the function is e.g. instantiated for every argument to ``class_<>``. So that's the whole reason that the ``type_caster_generic`` exists: to factor out the non-type-dependent parts (which is actually a fair bit of code) so that they generate only 1 symbol and 1 implementation.

wjakob commented:
Here's a general thought: any kind of access to dtype members from Python is going to be pretty slow! Probably not the way that you would write performance critical code, which should use vectorization.

If it's not performance critical, that also means that the pybind11 implementation could be simpler and tolerate some slowdowns. The main issue I see with the current code in the PR is that it not only adds extra map data structures, but that it touches the casting code for *everything*.

Perhaps the simpler way to go is to use the existing implicit cast feature with some kind of helper function that can be called to manually enable specific kinds of dtypes that can be accessed in this way.

aldanor commented:
@wjakob I've implemented the second option, now it has a minimal footprint, no more than it would be if an implicit conversion would be used, both code-bloat-wise and performance-wise.

// I've also cleaned up things a bit so `detail::get_type_info()` is used wherever possible instead of doing `registered_types_cpp.find()` all over the place.

aldanor commented:
// Btw I think this branch should also work with `sequence_t` and `iterable_t` from the typed-iterators PR, making it possible to pass numpy arrays directly where an iterable/sequence of a structured type is expected. I could add a test for that when/after it's merged.

wjakob commented on include/pybind11/cast.h:
```diff
@@ -231,22 +237,19 @@ class type_caster_generic {
         auto &internals = get_internals();
 
-        auto it = internals.registered_types_cpp.find(std::type_index(*type_info));
-        if (it == internals.registered_types_cpp.end()) {
```
Although less nice from a coding style view, I think that it would be nicer to leave the longer ``internals.registered_types_cpp.find`` calls here. This is a very "hot" piece of code, and the latter are more efficient since they don't have to resolve the ``internals`` data structure multiple times.

aldanor replied:
I think this one's actually the only code block where this kind of hackery is justified because it *is* the hot path, being the generic type caster. Mind if just revert this one? 

(another you didn't mention, the implicit conversions one, is also another "cold" path, likely to be just called once in the initialization, so should be cool)

wjakob commented on include/pybind11/pybind11.h:
```diff
@@ -610,8 +607,7 @@ class generic_type : public object {
 
-        if (internals.registered_types_cpp.find(tindex) !=
-            internals.registered_types_cpp.end())
+        if (get_type_info(*(rec->type)))
```
same here.

aldanor replied:
Same here as above, this is actually another freezing-cold code path :)

wjakob commented on include/pybind11/pybind11.h:
```diff
@@ -216,9 +214,8 @@ class cpp_function : public function {
                 const std::type_info *t = types[type_index++];
                 if (!t)
                     pybind11_fail("Internal error while parsing type signature (1)");
-                auto it = registered_types.find(std::type_index(*t));
```
same here.

aldanor replied:
Btw, taking a more detailed look -- this is a completely "cold" piece of code though, likely to only be called once per module initialization? (`cpp_function` setup)

wjakob replied:
Indeed, that one is. I remember seeing something in ``type_caster_generic`` though.

wjakob commented on include/pybind11/pybind11.h:
```diff
@@ -1589,11 +1586,8 @@ inline function get_type_overload(const void *this_ptr, const detail::type_info
-    if (it == cpp_types.end())
-        return function();
-    return get_type_overload(this_ptr, (const detail::type_info *) it->second, name);
+    auto tinfo = detail::get_type_info(typeid(T));
```
The change involving ``get_type_info()`` looks reasonable here.

wjakob commented on include/pybind11/common.h:
```diff
@@ -321,6 +321,7 @@ struct internals {
     std::unordered_map<const void *, void*> registered_types_py;       // PyTypeObject* -> type_info
     std::unordered_multimap<const void *, void*> registered_instances; // void * -> PyObject*
     std::unordered_set<std::pair<const PyObject *, const char *>, overload_hash> inactive_overload_cache;
+    std::unordered_map<std::type_index, std::vector<bool (*)(PyObject *, void *&)>> direct_conversions;
```
It wasn't clear to me, why this unordered map was needed, as compared to storing the ``std::vector<>`` (and not a pointer to it) directly in the ``py::detail::type_info``.

wjakob replied:
For instance, the generic_caster won't be able to deal with a dtype that has not have a proper ``py::detail::type_info`` record, so it seems unnecessary to track the associated conversions with a new global data structure.

wjakob replied:
The only potential issue I see is that ``register_dtype`` is called before the associated ``class_<>`` constructor has executed. Like with inheritance and similar features that depend on some prior initialization, I think that it's fair to impose a mandatory ordering here to keep the library internals simple.

aldanor commented:
Re: caching internals, it's just one check:

```cpp
if (internals_ptr)
    return *internals_ptr;
```

The rest of the code where it's used is likely to have more overhead than a single is-null check :) However if you'd like so I'll revert those parts (or add an overload of `get_type_info()`) that takes `internals` as argument as well -- I'm convinced it'd be nice if library code didn't touch `registered_types_cpp` directly if humanly possible).

Re: direct conversions being stored as a pointer in the typeinfo and the actual converters in a separate map - is because this way it doesn't care if `PYBIND11_NUMPY_DTYPE` or `py::class_<>` was called first, since the converters are guaranteed to be auto-magically shared between the two. The only price you pay, really, is a single pointer dereference when it comes to direct conversions. The difference between this case and e.g. implicit conversions is that in this case no error will be thrown if the order of declarations was wrong (i.e. typeinfo didn't exist when the numpy dtype was bound) -- it will just silently imply that direct converters for this dtype won't work -- in other cases I believe you will get an error which is easier to deal with.

wjakob commented:
> Re: caching internals, it's just one check

Right -- but it's a function that can't be inlined (intentionally), so this prevents a function call in a hot code path from being optimized away. I would prefer simply removing that part of the patch -- all of this is deep library-internal code, i.e. it's fine to mess with internal library data structures ;).

Regarding the second point: this is fine with me then.

aldanor commented:
Done, reverted the lookups in generic type caster, the rest is pretty much all initialization code.

wjakob commented:
Ok, merging then. It would be super-nice to have documentation for the all the fancy new NumPy features at some point :)

aldanor commented:
Agreed, definitely on my todo list :)

wjakob commented:
sounds good to me!
> On Oct 25, 2016, at 12:51 AM, Ivan Smirnov <notifications@github.com> wrote:
> 
> @aldanor commented on this pull request.
> 
> In include/pybind11/cast.h <https://github.com/pybind/pybind11/pull/453>:
> 
> > @@ -231,22 +237,19 @@ class type_caster_generic {
>  
>          auto &internals = get_internals();
>  
> -        auto it = internals.registered_types_cpp.find(std::type_index(*type_info));
> -        if (it == internals.registered_types_cpp.end()) {
> I think this one's actually the only code block where this kind of hackery is justified because it is the hot path, being the generic type caster. Mind if just revert this one?
> 
> (another you didn't mention, the implicit conversions one, is also another "cold" path, likely to be just called once in the initialization, so should be cool)
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/pybind/pybind11/pull/453>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ABJdrZcAy0efGkVA7VhKez5l5LzI8KFGks5q3TaHgaJpZM4KcTnc>.
> 
> {"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/pybind/pybind11","title":"pybind/pybind11","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/pybind/pybind11"}},"updates":{"snippets":[{"icon":"PERSON","message":"@aldanor commented on #453"}],"action":{"name":"View Pull Request","url":"https://github.com/pybind/pybind11/pull/453"}}}
